pandas/core/indexing.py

```diff
@@ -2407,29 +2407,54 @@ def convert_to_index_sliceable(obj, key):
     return None
 
 
-def check_bool_indexer(ax, key):
-    # boolean indexing, need to check that the data are aligned, otherwise
-    # disallowed
+def check_bool_indexer(index: Index, key) -> np.ndarray:
+    """
+    Check if key is a valid boolean indexer for an object with such index and
+    perform reindexing or conversion if needed.
+
+    This function assumes that is_bool_indexer(key) == True.
+
+    Parameters
+    ----------
+    index : Index
+        Index of the object on which the indexing is done
+    key : list-like
+        Boolean indexer to check
 
-    # this function assumes that is_bool_indexer(key) == True
+    Returns
+    -------
+    result: np.array
+        Resulting key
 
+    Raises
+    ------
+    ValueError
+        If the key does not have the same length as index
+
+    IndexingError
+        If the index of the key is unalignable to index
+
+    """
     result = key
-    if isinstance(key, ABCSeries) and not key.index.equals(ax):
-        result = result.reindex(ax)
+    if isinstance(key, ABCSeries) and not key.index.equals(index):
+        result = result.reindex(index)
         mask = isna(result._values)
         if mask.any():
             raise IndexingError('Unalignable boolean Series provided as '
                                 'indexer (index of the boolean Series and of '
-                                'the indexed object do not match')
+                                'the indexed object do not match).')
         result = result.astype(bool)._values
-    elif is_sparse(result):
-        result = result.to_dense()
-        result = np.asarray(result, dtype=bool)
     else:
-        # is_bool_indexer has already checked for nulls in the case of an
-        # object array key, so no check needed here
+        if is_sparse(result):
+            result = result.to_dense()
         result = np.asarray(result, dtype=bool)
 
+        # GH26658
+        if len(result) != len(index):
```

Comment:
I'm pretty confused about this check. How is this solving the original issue? I may be missing something.

Reply:
The way I understood the issue is that the code is not doing what the documentation says, as linked by yourself in the issue description:

> You may select rows from a DataFrame using a boolean vector the same length as the DataFrame’s index

So, if the lengths are different, an exception gets raised. Do you think it should be handled in a different way?

Reply:
Sorry, I missed the `result = key` assignment at the start.